### ISSUE TITLE
Add Entropic Retraining

### ISSUE THREAD

**Issue by philipsperl** [fra31/auto-attack#41]
**Paper**: [Optimizing Information Loss Towards Robust Neural Networks](https://arxiv.org/abs/2008.03072)

**Venue**: [DYNAMICS 2020 – DYnamic and Novel Advances in Machine Learning and Intelligent Cyber Security](https://www.acsac.org/2020/workshops/dynamics/)

**Dataset and threat model**: CIFAR-10, Linf, eps = 8/255

**Code**: Not published

**Pre-trained model**: Pre-trained model available [here](https://owncloud.fraunhofer.de/index.php/s/05oQYKNbZ5Bm2q0)

**Log file**: Log file of the evaluation available [here](https://owncloud.fraunhofer.de/index.php/s/MCbPyYsgte4CcYX)

**Additional data**: No

**Clean and robust accuracy**: Clean accuracy: 81.75%. Robust accuracy: 41.88%

**Architecture**: Simple convolutional network: Four convolutional layers, the first two with filter of size 32, the second pair with filter size of 64. Each pair of convolutional layers is followed by a max-pooling layer. Finally, one flatten layer and six dense layers with 512, 256, 128, 128, 84, and 10 neurons each. Script with the definition of the model available [here](https://owncloud.fraunhofer.de/index.php/s/ewpRUdHnwS8ib1F).

**Description of the model/defense**: New loss: During training, which is based normal data only, an adapted loss function is used. The additional loss-term is based on information theoretic inspired metrics. The method does not require the generation of adversarial examples during training.

**Comment by fra31** [comment]
Hi,

thanks for the submission! Could it be possible to include a script with the definition of the model for easier restoring?

**Comment by philipsperl** [comment]
Hi, thank you for considering our submission. We added a link to a script with the definition to the model.

**Comment by fra31** [comment]
Thanks for sharing the script. I noticed that the model definition includes the application of the softmax: while this doesn't change the predictions, it might be problematic since the attacks expect logits (before softmax). Then I set `activation=None` instead in the final layer.

When running AA on the model, I got (for Linf, `eps = 8/255`)

clean accuracy: 81.75%
robust accuracy 0.00%.

However, the robust accuracy with only APGD on the cross-entropy is 47.48%. I think the cause of this behavior is that for many points the logit of the correct class is the only positive one, but all have absolute values in the order of 10^2 or 10^3. This makes the computation of the gradients problematic if the softmax function is applied (as with the cross-entropy loss).

Please let me know if you can confirm these results or I missed something.

**Comment by philipsperl** [comment]
Thank you for your evaluation and the very helpful information. Under the settings you mentioned, I can reproduce your results. For further evaluations I might need to retrain the model, as the applied changes affect the behaviour of our approach. I'll provide an updated model and evaluation as soon as possible.
